ktor-client/ktor-client-android/src/io/ktor/client/engine/android/AndroidClientEngine.kt

```diff
@@ -60,40 +66,43 @@ open class AndroidClientEngine(override val config: AndroidEngineConfig) : HttpC
                 contentLength?.let { setFixedLengthStreamingMode(it.toInt()) } ?: setChunkedStreamingMode(0)
                 doOutput = true
 
-                outgoingContent.writeTo(outputStream)
+                outgoingContent.writeTo(outputStream, callContext)
             }
         }
 
         connection.connect()
-        val content = connection.content(dispatcher)
+        val content = connection.content(callContext)
         val headerFields = connection.headerFields
 
         val responseHeaders = HeadersBuilder().apply {
             headerFields?.forEach { (key, values) -> key?.let { appendAll(it, values) } }
         }.build()
 
         return AndroidHttpResponse(
-            call, content, context,
+            call, content,
             responseHeaders, requestTime, GMTDate(),
             HttpStatusCode.fromValue(connection.responseCode), HttpProtocolVersion.HTTP_1_1,
+            callContext,
             connection
         )
     }
 }
 
-internal fun OutgoingContent.writeTo(stream: OutputStream): Unit = stream.use {
+internal fun OutgoingContent.writeTo(
+    stream: OutputStream, callContext: CoroutineContext
+): Unit = stream.use {
     when (this) {
         is OutgoingContent.ByteArrayContent -> it.write(bytes())
-        is OutgoingContent.ReadChannelContent -> readFrom().toInputStream().copyTo(it)
+        is OutgoingContent.ReadChannelContent -> readFrom().toInputStream(callContext[Job]).copyTo(it)
         is OutgoingContent.WriteChannelContent -> {
-            writer(Unconfined) { writeTo(channel) }.channel.toInputStream().copyTo(it)
+            GlobalScope.writer(callContext) { writeTo(channel) }.channel.toInputStream(callContext[Job]).copyTo(it)
```

Comment:
Now we have these coroutines running on dispatcher instead of unconfined 

```kotlin
override val coroutineContext: CoroutineContext by lazy {
        dispatcher + Supervisor() + CoroutineName("$engineName-context")
    }

    protected fun createCallContext() = coroutineContext + CompletableDeferred<Unit>()
```

Reply:
Yep. And it's absolutely fine: we should avoid to use Unconfined on the Android(network tasks should avoid UI thread).